oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/VersionGarbageCollector.java

```diff
@@ -1246,6 +1349,256 @@ private void collectRevisionsOlderThan24hAndBetweenCheckpoints(final NodeDocumen
+                // child nodes could be having _commitRoots pointing to this doc
+                // and this revision - in which case it wouldn't be orphaned.
+                // without scanning the children (or having knowledge about
+                // which revisions are referenced by children) it is unsafe
+                // to delete committed revisions.
+                // if they are uncommitted, then they are garbage, as then they
+                // are either normal commits (thus a not finished commit)
+                // or an unmerged branch commit. In both cases they are garbage.
+                boolean isCommitted = isCommitted(nodeStore.getCommitValue(revision, doc));
+                if (isCommitted) {
+                    if (isRoot) {
+                        // root : cannot remove since it could be referenced and required by a child
+                        // also add to allRequiredRevs
+                        allRequiredRevs.add(revision);
+                        continue;
+                    } else if (!isBC) {
+                        // non root and normal : cannot remove, same as above,
+                        // it could be referenced and required by a child 
+                        // also add to allRequiredRevs
+                        allRequiredRevs.add(revision);
+                        continue;
+                    } else {
+                        // non root and bc : can remove, as non root bc cannot be referenced by child
+                    }
+                }
+                Operation has = updateOp.getChanges().get(new Key(NodeDocument.REVISIONS, revision));
+                if (has != null) {
+                    // then skip
+                    continue;
+                }
+                NodeDocument.removeRevision(updateOp, revision);
+                deletedRevsCount++;
+            }
+            // "_commitRoot"
+            for (Revision revision : doc.getLocalCommitRoot().keySet()) {
+                if (!allRequiredRevs.contains(revision)) {
+                    Operation has = updateOp.getChanges().get(new Key(NodeDocument.COMMIT_ROOT, revision));
+                    if (has != null) {
+                        // then skip
+                        continue;
+                    }
+                    NodeDocument.removeCommitRoot(updateOp, revision);
+                    deletedRevsCount++;
+                }
+            }
+            // "_bc"
+            for (Revision revision : doc.getLocalBranchCommits()) {
+                if (!allRequiredRevs.contains(revision)) {
+                    Operation has = updateOp.getChanges().get(new Key(NodeDocument.BRANCH_COMMITS, revision));
+                    if (has != null) {
+                        // then skip
+                        continue;
+                    }
+
+                    NodeDocument.removeBranchCommit(updateOp, revision);
+                    deletedRevsCount++;
+                }
+            }
```

Comment:
```suggestion

deletedRevsCount += getDeletedRevsCount(doc.getLocalBranchCommits(), updateOp, allRequiredRevs, BRANCH_COMMITS, NodeDocument::removeBranchCommit);
return deletedRevsCount;
}

            private int getDeletedRevsCount(Set<Revision> revisionSet, UpdateOp updateOp, Set<Revision> allRequiredRevs, String updateOpKey, BiConsumer<UpdateOp, Revision> op) {

            return revisionSet.stream().filter(r -> !allRequiredRevs.contains(r))
                    .mapToInt(r -> {
                        Operation has = updateOp.getChanges().get(new Key(updateOpKey, r));
                        if (has != null) {
                            // then skip
                            return 0;
                        }

                        op.accept(updateOp, r);
                        return 1;
                    }).sum();
        }
```

Reply:
incorporated - and fixed brackets in https://github.com/apache/jackrabbit-oak/pull/1372/commits/4bbb8ed82c9ac7ae53994f9de1cbed15ca515718